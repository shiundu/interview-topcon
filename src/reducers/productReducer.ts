import { Product } from '../models';
import { Action } from '../actions/products';
import { ActionTypes } from '../actions/types';

export interface State {
    products: Product[]
};

// Define our initialState
export const initialState: State = {
    products: []
};

export function reducer(state: State = initialState, action: Action) {
    switch (action.type) {

        case ActionTypes.INITIAL_PRODUCT_LIST: {
            const products = action.payload.products.reverse();
    
            return {
                ...state,
                products
            }
        }

        case ActionTypes.ADD_PRODUCT: {
            let product = action.payload.product;
            product.id = state.products.length  + 1;
            product.imgUrl = product.imgUrl.length > 0 ? product.imgUrl : 'https://via.placeholder.com/200'

            return {
                ...state,
                products: [product, ...state.products]
            }
        }

        case ActionTypes.UPDATE_PRODUCT: {
            const prod = action.payload.product
            return {
                ...state,
                products: state.products.map(product => {
                    if (product.id === prod.id) {
                        product.name = prod.name.length > 0 ? prod.name : prod.name;
                        product.price = prod.price.length > 0 ? prod.price : product.price;
                        product.imgUrl = prod.imgUrl.length > 0 ? product.imgUrl : 'https://via.placeholder.com/200'
                        return product
                    }
                    return product;
                })
            }
        }

        case ActionTypes.DELETE_PRODUCT: {
            const { productId } = action.payload
            return {
                ...state,
                products: state.products.filter(product => product.id !== productId)
            }
        }

        default:
            return state
    }
}