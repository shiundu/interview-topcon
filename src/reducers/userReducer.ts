import { User } from '../models';
import { Action } from '../actions/auth';
import { ActionTypes } from '../actions/types';

// Define the State interface for user reducer
export interface State {
    user: User
}

// initialState
export const initialState: State = {
    user: {
        id: 0,
        firstName: '',
        lastName: '',
        email: '',
        authenticationToken: '',
        error: false
    }
}

export function reducer(state: State = initialState, action: Action) {
    switch (action.type) {

        case ActionTypes.LOGIN_USER: {
            const user = action.payload.user
    
            return {
            ...state,
            user 
            }
        }

        case ActionTypes.LOGOUT_USER: {
            return {
                ...state,
                ...initialState
            }
        }

        default:
        return state
    }
};