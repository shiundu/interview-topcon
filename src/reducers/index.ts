import { State } from './index';
import { combineReducers } from 'redux';
import * as fromProducts from './productReducer';
import * as fromUser from './userReducer';

export interface State {
  products:  fromProducts.State,
  user: fromUser.State
}

export const initialState: State = {
  products: fromProducts.initialState,
  user: fromUser.initialState
}

export const reducer = combineReducers<State>({
  products: fromProducts.reducer,
  user: fromUser.reducer
})