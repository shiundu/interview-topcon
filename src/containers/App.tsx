import { connect } from 'react-redux';
import { State } from '../reducers';
import { getUser } from '../selectors';
import { logoutUser } from '../actions';
import AddProduct from './AddProduct';
import Products from './Products';
import Login from './Login';
import withHeader from '../components/bundlerHOC';

const mapStateToProps = (state: State) => ({
    user: getUser(state)
})

const mapDispatchToProps = {
    logoutUser: logoutUser
}

export default connect(mapStateToProps, mapDispatchToProps)(withHeader(Login, AddProduct, Products));