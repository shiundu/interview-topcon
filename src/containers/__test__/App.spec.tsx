import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import App from '../App';
import { initialState } from '.././../reducers';

const mockStore = configureMockStore();
const store = mockStore({...initialState});

describe('App test', () => {
    it('App renders correctly', () => {
        const tree = renderer.create(
        <Provider store={store}>
              <App />
            </Provider>)
          .toJSON();

        expect(tree).toMatchSnapshot();
    });
})
