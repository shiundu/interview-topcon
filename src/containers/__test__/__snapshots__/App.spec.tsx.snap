// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App test App renders correctly 1`] = `
<div
  className="col"
>
  <form
    className="form-signin"
  >
    <h1
      className="h3 mb-3 font-weight-normal"
    >
      Please sign in
    </h1>
    <input
      className="form-control"
      id="inputEmail"
      name="username"
      onChange={[Function]}
      placeholder="Email address: test@test.co"
      type="email"
    />
    <input
      className="form-control"
      id="inputPassword"
      name="password"
      onChange={[Function]}
      placeholder="Password: password"
      type="password"
    />
    <button
      className="btn btn-lg btn-primary btn-block"
      onClick={[Function]}
    >
      Sign in
    </button>
    <p
      className="mt-5 mb-3 text-muted"
    >
      Â© 2017-
    </p>
  </form>
</div>
`;
